drivers/flash/flash_stm32_qspi.c

```diff
@@ -0,0 +1,911 @@
+
+	if (magic != JESD216_SFDP_MAGIC) {
+		LOG_ERR("SFDP magic %08x invalid", magic);
+		return -EINVAL;
+	}
+
+	LOG_INF("%s: SFDP v %u.%u AP %x with %u PH", dev->name,
+		hp->rev_major, hp->rev_minor, hp->access, hp->nph);
+
+	const struct jesd216_param_header *php = hp->phdr;
+	const struct jesd216_param_header *phpe = php + MIN(decl_nph, hp->nph);
+
+	while (php != phpe) {
+		uint16_t id = jesd216_param_id(php);
+
+		LOG_INF("PH%u: %04x rev %u.%u: %u DW @ %x",
+			(php - hp->phdr), id, php->rev_major, php->rev_minor,
+			php->len_dw, jesd216_param_addr(php));
+
+		if (id == JESD216_SFDP_PARAM_ID_BFP) {
+			union {
+				uint32_t dw[MIN(php->len_dw, 20)];
+				struct jesd216_bfp bfp;
+			} u;
+			const struct jesd216_bfp *bfp = &u.bfp;
+
+			ret = qspi_read_sfdp(dev, jesd216_param_addr(php),
+					     (uint8_t *)u.dw, sizeof(u.dw));
+			if (ret == 0) {
+				ret = spi_nor_process_bfp(dev, php, bfp);
+			}
+
+			if (ret != 0) {
+				LOG_ERR("SFDP BFP failed: %d", ret);
+				break;
+			}
+		}
+		++php;
+	}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	ret = setup_pages_layout(dev);
+	if (ret != 0) {
+		LOG_ERR("layout setup failed: %d", ret);
+		return -ENODEV;
+	}
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+	LOG_INF("Device %s initialized", DEV_NAME(dev));
+
+	return 0;
+}
+
+#define DMA_CHANNEL_CONFIG(node, dir)					\
+		DT_DMAS_CELL_BY_NAME(node, dir, channel_config)
+
+#define QSPI_DMA_CHANNEL_INIT(node, dir)				\
+	.name = DT_DMAS_LABEL_BY_NAME(node, dir),			\
+	.channel = DT_DMAS_CELL_BY_NAME(node, dir, channel),		\
+	.reg = (DMA_TypeDef *)DT_REG_ADDR(				\
+				   DT_PHANDLE_BY_NAME(node, dmas, dir)),\
+	.cfg = {							\
+		.dma_slot = DT_DMAS_CELL_BY_NAME(node, dir, slot),	\
+		.source_data_size = STM32_DMA_CONFIG_PERIPHERAL_DATA_SIZE( \
+					DMA_CHANNEL_CONFIG(node, dir)), \
+		.dest_data_size = STM32_DMA_CONFIG_MEMORY_DATA_SIZE(    \
+					DMA_CHANNEL_CONFIG(node, dir)), \
+		.channel_priority = STM32_DMA_CONFIG_PRIORITY(		\
+					DMA_CHANNEL_CONFIG(node, dir)), \
+		.dma_callback = qspi_dma_callback,			\
+	},								\
+
+#define QSPI_DMA_CHANNEL(node, dir)					\
+	.dma = {							\
+		COND_CODE_1(DT_DMAS_HAS_NAME(node, dir),		\
+			(QSPI_DMA_CHANNEL_INIT(node, dir)),		\
+			(NULL))						\
+		},
+
+#define QSPI_FLASH_MODULE(drv_id, flash_id) 				\
+		(DT_DRV_INST(drv_id), qspi_nor_flash_##flash_id)
+
+static void flash_stm32_qspi_irq_config_func(const struct device *dev);
+
+static const struct soc_gpio_pinctrl qspi_pins[] =
+					ST_STM32_DT_PINCTRL(quadspi, 0);
+
+#define STM32_QSPI_NODE DT_PARENT(DT_DRV_INST(0))
+
+static const struct flash_stm32_qspi_config flash_stm32_qspi_cfg = {
+	.regs = (QUADSPI_TypeDef *)DT_REG_ADDR(STM32_QSPI_NODE),
+	.pclken = {
+		.enr = DT_CLOCKS_CELL(STM32_QSPI_NODE, bits),
+		.bus = DT_CLOCKS_CELL(STM32_QSPI_NODE, bus)
+	},
+	.irq_config = flash_stm32_qspi_irq_config_func,
+	.flash_size = DT_INST_PROP(0, size) / 8U,
+	.max_frequency = DT_INST_PROP(0, spi_max_frequency),
```

Comment:
Can there ever be more than one instance?

Reply:
I don't think so, but from what I understand from @galak the `DT_INST_` version is the preferred one when possible.
At least I spared the iteration macro stuff.